dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/orgunit/OrgUnitMergeValidator.java

```diff
@@ -0,0 +1,87 @@
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.merge.orgunit;
+
+import org.hisp.dhis.common.IllegalQueryException;
+import org.hisp.dhis.feedback.ErrorCode;
+import org.hisp.dhis.feedback.ErrorMessage;
+import org.springframework.stereotype.Service;
+
+/**
+ * Validation service for org unit merge requests.
+ *
+ * @author Lars Helge Overland
+ */
+@Service
+public class OrgUnitMergeValidator
+{
+    /**
+     * Validates the given {@link OrgUnitMergeRequest}. Throws
+     * {@link IllegalQueryException} if validation fails.
+     *
+     * @param request the {@link OrgUnitMergeRequest}.
+     * @throws IllegalQueryException if validation failed.
+     */
+    public void validate( OrgUnitMergeRequest request )
+        throws IllegalQueryException
+    {
+        ErrorMessage error = validateForErrorMessage( request );
+
+        if ( error != null )
+        {
+            throw new IllegalQueryException( error );
+        }
+    }
+
+    /**
+     * Validates the given {@link OrgUnitMergeRequest}.
+     *
+     * @param request the {@link OrgUnitMergeRequest}.
+     * @return an {@link ErrorMessage} if the validation failed, or null if
+     *         validation was successful.
+     */
+    public ErrorMessage validateForErrorMessage( OrgUnitMergeRequest request )
+    {
+        ErrorMessage error = null;
+
+        if ( request.getSources().isEmpty() )
+        {
+            error = new ErrorMessage( ErrorCode.E1500 );
+        }
+        else if ( request.getTarget() == null )
+        {
+            error = new ErrorMessage( ErrorCode.E1501 );
+        }
+        else if ( request.getSources().contains( request.getTarget() ) )
+        {
+            error = new ErrorMessage( ErrorCode.E1502 );
+        }
+
+        return error;
```

Comment:
Seen this several times and would think that flow is much more clear (especially when this grows longer) when using a `return new ErrorMessage(...)` in each if block.

Reply:
There are people who argue multiple exit points in a method is bad practice too, I think this is subject to opinion :-)

Reply:
Sure. Just recommending what I would consider lower mental load.

Introducing a variable is as the name suggests introduces mutable state that our brain now needs to track. Also we need to read all the code of the method to make sure each branch is actually only setting that variable once at which point we build a mental model which is much closer to multi-return as we "exit" our considerations at the first branch we enter because we concluded that this effectively will be the last time something happens to that place we have made a burden for us to keep track of :grin:

Reply:
Sure updated it.